README.rst
.. image:: https://travis-ci.org/Luqqk/catchments.svg?branch=master
    :target: https://travis-ci.org/Luqqk/catchments

.. image:: https://coveralls.io/repos/github/Luqqk/catchments/badge.svg
    :target: https://coveralls.io/github/Luqqk/catchments

🌍 catchments
=============

Python wrapper for multiple APIs, that provide catchments-areas.
It allows to acquire and manipulate catchments from those APIs.
It can help you answer the questions like: "Where can I get in 30 min from that point?"
taking into account e.g. real traffic, transport type, specified range. 

.. image:: img/catchments.png
    :height: 400px
    :width: 400px
    :align: center
    :target: https://github.com/Luqqk/catchments/blob/master/img/catchments.png

Installation
------------

.. code-block:: bash

    $ pip install catchments

Usage
-----

Currently there are implemented two classes: **SkobblerAPI("api_key")** and **HereAPI("app_id", "app_code")**.

You can use them as follows:

.. code-block:: python

    >>> from catchments import SkobblerAPI

    >>> # get catchment from Skobbler API
    >>> skobbler = SkobblerAPI('your_api_key')
    >>> # if you don't provide params values default ones will be used
    >>> params = {"range": 600, "highways": 1}
    >>> catchment = skobbler.get_catchment({"lat": 52.40, "lon": 16.93}, **params)
    >>> {"realReach": {...} ...}
    >>> geojson = skobbler.catchment_as_geojson(catchment)
    >>> {"type": "Feature", geometry: {"type": "Polygon", ...}, ...}
    >>> skobbler.save_as_geojson(geojson)
    >>> 'SKOBBLER_52.40_16.93.geojson'

As you can see **.get_catchment** method uses **params** as second argument. Params keys names should be exactly the same
as mentioned in APIs documentations, otherwise they will be ignored and default values will be used.

Params supported by **SKOBBLER** and **HERE**:

`SKOBBLER <https://developer.skobbler.com/getting-started/web#sec3>`_ (startMercator, response_type - not supported)

`HERE <https://developer.here.com/rest-apis/documentation/routing/topics/request-isoline.html>`_

You can use also inbuilt command line scripts which accept \*.csv file input with points as coordinates resource.
Scripts generate \*.geojson files for every point in given \*.csv file.

Example \*.csv file structure (name column is optional):

+------------+------------+------------+ 
|    name    |    lat     |    lon     | 
+============+============+============+ 
|   point1   |  52.0557   |  16.8278   | 
+------------+------------+------------+ 
|   point2   |  52.4639   |  16.9410   | 
+------------+------------+------------+ 

Example command line script usage:

.. code-block:: bash

    $ catchments-skobbler.py -k your_api_key -p path/to/file/with/points/*.csv

All scripts and their options are mentioned below:

.. code-block:: bash

    $ catchments-skobbler.py

* -k --key [REQUIRED] [DEFAULT: **None**]

* -p --points [REQUIRED] [DEFAULT: **None**]

* -r --range - [OPTIONAL] [DEFAULT: **600**]

* -u --units - [OPTIONAL] [DEFAULT: **sec**]

* -t --transport - [OPTIONAL] [DEFAULT: **car**]

* -l --toll - [OPTIONAL] [DEFAULT: **0**]

* -w --highways - [OPTIONAL] [DEFAULT: **0**]

* -n --nonReachable - [OPTIONAL] [DEFAULT: **0**]

.. code-block:: bash

    $ catchments-here.py

* -i --app_id [REQUIRED] [DEFAULT: **None**]

* -c --app_code [REQUIRED] [DEFAULT: **None**]

* -r --range - [OPTIONAL] [DEFAULT: **600**]

* -e --range-type - [OPTIONAL] [DEFAULT: **time**]

* -m --mode - [OPTIONAL] [DEFAULT: **fastest;car;traffic:disabled**]

Tests
-----

.. code-block:: bash

    $ python setup.py test

TODO
----

* Add support for Mapzen API catchments
* Add support for OpenRouteService catchments
* Add support for concurrent HTTP requests


HISTORY.rst
.. :changelog:

Release History
---------------

1.1.1 (2017-05-04)
++++++++++++++++++

* Split catchments-cls.py into separate scripts (one for each API)
* Documentation

1.0.0 (2017-05-01)
++++++++++++++++++

* Bug fixes, code reorganization
* Better documentation

0.9.1b (2017-03-11)
+++++++++++++++++++

* Required params validation in catchments-cls.py

0.9.0b (2017-03-11)
+++++++++++++++++++

* Initial release (beta)


catchments/tests/test_data.py
EXAMPLE_SKOBBLER_PARAMS = {"range": 800}

EXAMPLE_HERE_PARAMS = {"range": 1200}

EXAMPLE_SKOBBLER_CATCHMENT = {
    "realReach": {
        "gpsBBox": [10.00,45.00,16.00,52.00],
        "gpsPoints": [
            -180.0,85.051129,
            180.0,85.051129,
            180.0,-85.051129,
            -180.0,-85.051129,
            10.10,45.20,
            11.10,46.20,
            12.10,47.20,
            13.10,48.20,
            14.10,49.20,
            15.10,50.20
        ]
    },
    "name": "test_point"
}

EXAMPLE_HERE_CATCHMENT = {
    "response": {
        "isoline": [
            {"component": [
                    {"shape": [
                            "50.00,16.00",
                            "50.10,16.10",
                            "50.20,16.20",
                            "50.30,16.30",
                            "50.00,16.00",
                        ]
                    }
                ]
            }
        ]
    },
    "name": "test_point"
}

EXAMPLE_SKOBBLER_GEOJSON = {
    "type": "Feature",
    "geometry": {
        "type": "Polygon",
        "coordinates": [
            [
                [10.10,45.20],
                [11.10,46.20],
                [12.10,47.20],
                [13.10,48.20],
                [14.10,49.20],
                [15.10,50.20],
                [10.10,45.20]
            ]
        ]
    },
    "properties": {
        "name": "test_point"
    }
}

EXAMPLE_HERE_GEOJSON = {
    "type": "Feature",
    "geometry": {
        "type": "Polygon",
        "coordinates": [
            [
                [16.00,50.00],
                [16.10,50.10],
                [16.20,50.20],
                [16.30,50.30],
                [16.00,50.00]
            ]
        ]
    },
    "properties": {
        "name": "test_point"
    }
}


catchments/tests/test_here.py
from unittest import TestCase
from unittest.mock import patch, Mock
from io import StringIO
from tempfile import mkdtemp
from shutil import rmtree
from catchments import HereAPI
from .test_data import EXAMPLE_HERE_PARAMS, EXAMPLE_HERE_CATCHMENT, \
    EXAMPLE_HERE_GEOJSON
import os
import csv
import requests
# csv.OrderedDict is supported only in Python > 3.6
# collections.OrderedDict for backward compatibility (Python < 3.6)
import collections


# Run tests with:
# coverage run --branch --source=catchments/ setup.py test
# To check coverage report (with missing lines)
# coverage report -m


class TestRequestHereCatchment(TestCase):

    def setUp(self):
        self.here_api = HereAPI('app_id', 'app_code')
        self.here_point = {'lat': 50.0, 'lon': 16.0}
        self.here_mock_response = Mock()

    @patch('requests.get')
    def test_request_here_catchment(self, mock_request):
        successful_here_response = {
            "response": {
                "isoline": [
                    {"component": [
                            {"shape": []}
                        ]
                    }
                ]
            }
        }
        
        self.here_mock_response.json.return_value = successful_here_response
        mock_request.return_value = self.here_mock_response

        self.assertEqual(
            self.here_api.get_catchment(self.here_point, **EXAMPLE_HERE_PARAMS),
            successful_here_response
        )

    @patch('requests.get')
    def test_request_here_catchment_http_error(self, mock_request):
        skobbler_http_error_response = {}

        http_error = requests.exceptions.HTTPError()
        
        self.here_mock_response.json.return_value = skobbler_http_error_response
        self.here_mock_response.raise_for_status.side_effect = http_error
        mock_request.return_value = self.here_mock_response
        self.assertEqual(self.here_api.get_catchment(self.here_point, **EXAMPLE_HERE_PARAMS), None)


class TestHereCatchmentAsGeojson(TestCase):

    def setUp(self):
        self.here_api = HereAPI('app_id', 'app_code')
        self.here_geojson = EXAMPLE_HERE_GEOJSON
        self.here_geojson_feature = self.here_api.catchment_as_geojson(
            EXAMPLE_HERE_CATCHMENT
        )

    def test_here_catchment_as_geojson(self):
        self.assertEqual(self.here_geojson_feature, self.here_geojson)
    
    def test_here_invalid_api_response(self):
        invalid_here_response = {"response": {"type": "PermissionError"}}
        self.assertEqual(
            self.here_api.catchment_as_geojson(invalid_here_response),
            None
        )

class TestSaveAsGeojson(TestCase):

    def setUp(self):
        self.here_api = HereAPI('app_id', 'app_code')
        self.params = EXAMPLE_HERE_PARAMS
        # Create a temporary directory
        self.test_dir = mkdtemp()
        self.geojson = EXAMPLE_HERE_GEOJSON

    def tearDown(self):
        # Remove the directory after the test
        rmtree(self.test_dir)

    def test_save_as_geojson_with_save_in(self):
        path_to_save = self.here_api.save_as_geojson(self.geojson, save_in=self.test_dir)
        self.assertEqual(path_to_save, os.path.join(self.test_dir, 'HERE_test_point.geojson'))
    
    def test_save_as_geojson(self):
        path_to_save = self.here_api.save_as_geojson(self.geojson)
        os.remove(path_to_save)
        self.assertEqual(path_to_save, os.path.join(os.getcwd(), 'HERE_test_point.geojson'))


catchments/tests/test_utils.py
from unittest import TestCase
from io import StringIO
from tempfile import mkdtemp
from shutil import rmtree
from catchments.utils import load_input_data
import csv
# csv.OrderedDict is supported only in Python > 3.6
# collections.OrderedDict for backward compatibility (Python < 3.6)
import collections


# Run tests with:
# coverage run --branch --source=catchments/ setup.py test
# To check coverage report (with missing lines)
# coverage report -m


class TestLoadInputData(TestCase):

    def setUp(self):
        self.data_temp = StringIO()
        dict_writer = csv.DictWriter(self.data_temp, fieldnames=['lat', 'lon'])
        dict_writer.writeheader()
        dict_writer.writerow({'lat': 52.02, 'lon': 16.02})
        self.data_temp.seek(0)
    
    def test_load_csv(self):   
        data = load_input_data(self.data_temp)
        self.assertTrue(isinstance(data, csv.DictReader))

    def test_data_output(self):
        data = load_input_data(self.data_temp)
        for row in data:
            self.assertEqual(row, collections.OrderedDict([('lat', '52.02'), ('lon', '16.02')]))


skblerr.py
# !/usr/bin/python
# -*- coding: utf-8 -*-

from optparse import OptionParser
import os.path
import csv


def main():
    parser = OptionParser(usage='')
    parser.add_option('-r', '--range', type='int', help='')
    parser.add_option('-u', '--units', type='string', help='')
    parser.add_option('-t', '--transport', type='string', help='')
    parser.add_option('-f', '--file', type='string', help='')
    
    (options, args) = parser.parse_args()
    
    if not options.range:
        parser.error('Range is required')
    if not options.units:
        parser.error('Units are required [\'sec\', \'meter\']')
    if options.units not in ['sec', 'meter']:
        parser.error('Invalid units type [\'sec\', \'meter\']')
    if not options.transport:
        parser.error('Transport type is required [\'pedestrian\', \'bike\', \'car\']')
    if options.transport not in ['pedestrian', 'bike', 'car']:
        parser.error('Invalid units type [\'pedestrian\', \'bike\', \'car\']')
    if not options.file:
        parser.error('File is required (.csv with \',\' as delimiter)')
    if not os.path.isfile(options.file):
        parser.error('File doesn\'t exist.')
    
    print(options)
    print(args)


def read_file(file_path):
    with open(file_path) as csv_file:
        dialect = csv.Sniffer().sniff(csv_file.read(), delimiters=';,')
        csv_file.seek(0)
        return [row for row in csv.DictReader(csv_file, dialect=dialect)]

         
if __name__ == '__main__':
    main()

catchments/here.py
import os
import csv
import json
import requests


class HereAPI(object):
    """The HereAPI object implements HERE Isolines API."""

    def __init__(self, app_id, app_code):
        self.app_id = app_id
        self.app_code = app_code

    def _request(self, url, point, params):
        try:
            r = requests.get(url, params=params)
            r.raise_for_status()
        except requests.HTTPError:
            return None

        catchment = r.json()

        catchment['name'] = point.get(
            'name', '{}_{}'.format(point['lat'], point['lon'])
        )

        return catchment

    def get_catchment(self, point, **params):
        """Requests catchment from API provider.

        :param point (dictionary): {'name': 'place', 'lon': 50.0, 'lat': 20.0}
            'name' key is optional, 'lon' and 'lat' are required.

        :param params (**dictionary):
                supported keys:
                    mode, range, rangetype

        If optional params won't be supplied, default values will be used.

        Returns:
            API response if successful, None otherwise.
        """

        url = 'https://isoline.route.cit.api.here.com/routing/7.2/calculateisoline.json'

        request_params = {}

        request_params['start'] = 'geo!{1},{0}'.format(point['lon'], point['lat'])

        request_params['mode'] = params.get('mode', 'fastest;car;traffic:disabled')

        request_params['range'] = params.get('range', '600')

        request_params['rangetype'] = params.get('rangetype', 'time')

        request_params['app_id'] = self.app_id

        request_params['app_code'] = self.app_code

        return self._request(url, point, request_params)

    @staticmethod
    def catchment_as_geojson(catchment):
        """Processing catchment to GeoJSON format.

        :param catchment (dictionary)

        Returns:
            GeoJSON polygon feature if successful, None otherwise.
        """

        geojson = {
            "type": "Feature",
            "geometry": {
                "type": "Polygon", "coordinates": [[]]
            },
            "properties": {}
        }

        try:
            shape = catchment['response']['isoline'][0]['component'][0]['shape']
        except KeyError:
            return None

        coords = []
        for coord in shape:
            lat_lon = coord.split(',')
            coords.append(float(lat_lon[1]))
            coords.append(float(lat_lon[0]))

        for i, coord in enumerate(coords):
            if (i % 2 == 0):
                geojson['geometry']['coordinates'][0].append(
                    [coord, coords[i + 1]]
                )

        geojson['properties']['name'] = catchment['name']

        return geojson

    @staticmethod
    def save_as_geojson(geojson, save_in=None):
        """Save GeoJSON feature to *.geojson file.

        :param geojson (dictionary - GeoJSON feature)

        :param save_in (path)

        Returns:
           File with GeoJSON feature
           path_to_save: saved *.geojson file path
        """

        name = 'HERE_{}.geojson'.format(geojson['properties']['name'])

        if save_in:
            path_to_save = os.path.join(save_in, name)
        else:
            path_to_save = os.path.join(os.getcwd(), name)

        feature = json.dumps(geojson, indent=2)

        with open(path_to_save, 'w') as f:
            f.write(feature)

        return path_to_save


catchments/parsers.py
from optparse import OptionParser


def create_skobbler_parser():
    """Creates parser for SKOBBLER commandline arguments.

    Returns:
        parser (optparse.OptionParser)
    """

    parser = OptionParser()
    
    # Required parameters
    parser.add_option(
        '-k', '--key', type='string',
        help='SKOBBLER API key'
    )
    parser.add_option(
        '-p', '--points', type='string',
        help='*.csv file to read points from'
    )
    
    # Optional parameters
    parser.add_option(
        '-r', '--range', type='string', default='600',
        help='Range (int)'
    )
    parser.add_option(
        '-u', '--units', type='choice',
        choices=['sec', 'meter'], default='sec',
        help='(sec, meter)'
    )
    parser.add_option(
        '-t', '--transport', type='choice',
        choices=['pedestrian', 'bike', 'car'], default='car',
        help='(pedestrian, bike, car)'
    )
    parser.add_option(
        '-l', '--toll', type='choice',
        choices=['0', '1'], default='0',
        help='''Specifies whether to avoid or not 
        the use of toll roads in route calculation (0, 1)'''
    )
    parser.add_option(
        '-w', '--highways', type='choice',
        choices=['0', '1'], default='0',
        help='''Specifies whether to avoid or not
        the use of highways in route calculation (0, 1)'''
    )
    parser.add_option(
        '-n', '--nonReachable', type='choice',
        choices=['0', '1'], default='0',
        help='''Specifies whether to calculate
        or not the interior contours (non reachable areas)
        inside the RealReach™ (0, 1)'''
    )

    return parser


def create_here_parser():
    """Creates parser for HERE commandline arguments.

    Returns:
        parser (optparse.OptionParser)
    """

    parser = OptionParser()
    
    # Required parameters
    parser.add_option(
        '-i', '--app_id', type='string',
        help='HERE API app_id'
    )
    parser.add_option(
        '-c', '--app_code', type='string',
        help='HERE API app_code'
    )
    parser.add_option(
        '-p', '--points', type='string',
        help='*.csv file to read points from'
    )
    
    # Optional parameters
    parser.add_option(
        '-r', '--range', type='string', default='600',
        help='Range (int)'
    )
    parser.add_option(
        '-t', '--rangetype', type='string', default='time',
        help='(time, distance)'
    )
    parser.add_option(
        '-m', '--mode', type='string', default='fastest;car;traffic:disabled',
        help='''Mode - real time traffic and transport type
        (fastest;car;traffic:disabled)'''
    )

    return parser


catchments/skobbler.py
import os
import csv
import json
import requests


class SkobblerAPI(object):
    """The SkobblerAPI object implements Skobbler RealReach API."""

    def __init__(self, api_key):
        self.api_key = api_key

    def _request(self, url, point, params):
        try:
            r = requests.get(url, params=params)
            r.raise_for_status()
        except requests.HTTPError:
            return None

        catchment = r.json()

        catchment['name'] = point.get(
            'name', '{}_{}'.format(point['lat'], point['lon'])
        )

        return catchment

    def get_catchment(self, point, **params):
        """Requests catchment from API provider.

        :param point (dictionary):
            {'name': 'place', 'lon': 50.0, 'lat': 20.0}
            'name' key is optional, 'lon' and 'lat' are required.

        :param params (**dictionary):
                supported keys:
                    transport, range, units, toll, highways, non_reachable, jam

        If optional params won't be supplied, default values will be used.

        Returns:
            API response if successful, None otherwise.
        """

        url = 'http://{}.tor.skobbler.net/tor/RSngx/RealReach/json/20_5/en/{}'.format(
            self.api_key, self.api_key
        )

        request_params = {}

        request_params['start'] = '{1},{0}'.format(point['lon'], point['lat'])

        request_params['transport'] = params.get('transport', 'car')

        request_params['range'] = params.get('range', '600')

        request_params['units'] = params.get('units', 'sec')

        request_params['toll'] = params.get('toll', '0')

        request_params['highways'] = params.get('highways', '0')

        request_params['nonReachable'] = params.get('nonReachable', '0')

        request_params['response_type'] = 'gps'

        return self._request(url, point, request_params)

    @staticmethod
    def catchment_as_geojson(catchment):
        """Processing catchment to GeoJSON format.

        :param catchment (dictionary)

        Returns:
            GeoJSON polygon feature if successful, None otherwise.
        """

        geojson = {
            "type": "Feature",
            "geometry": {
                "type": "Polygon", "coordinates": [[]]
            },
            "properties": {}
        }

        try:
            coords = catchment['realReach']['gpsPoints']
            bbox = catchment['realReach']['gpsBBox']
        except KeyError:
            return None

        for i, coord in enumerate(coords):
            if (i % 2 == 0):
                if not (coord < bbox[0] or coord > bbox[2]):
                    geojson['geometry']['coordinates'][0].append(
                        [coord, coords[i + 1]]
                    )

        # Close GeoJSON polygon
        geojson['geometry']['coordinates'][0].append(
            geojson['geometry']['coordinates'][0][0]
        )
        geojson['properties']['name'] = catchment['name']

        return geojson

    @staticmethod
    def save_as_geojson(geojson, save_in=None):
        """Save GeoJSON feature to *.geojson file.

        :param geojson (dictionary - GeoJSON feature)

        :param save_in (path)

        Returns:
           File with GeoJSON feature
           path_to_save: saved *.geojson file path
        """

        name = 'SKOBBLER_{}.geojson'.format(geojson['properties']['name'])

        if save_in:
            path_to_save = os.path.join(save_in, name)
        else:
            path_to_save = os.path.join(os.getcwd(), name)

        feature = json.dumps(geojson, indent=2)

        with open(path_to_save, 'w') as f:
            f.write(feature)

        return path_to_save


catchments.py
# !/usr/bin/python
# -*- coding: utf-8 -*-

from optparse import OptionParser
import os.path
import csv
import requests
import json


def main():
    parser = OptionParser(usage='')
    # Required parameters
    parser.add_option('-a', '--api', type='string', help='')
    parser.add_option('-k', '--key', type='string', help='')
    parser.add_option('-f', '--file', type='string', help='')
    # Optional parameters
    parser.add_option('-r', '--range', type='int', default=600, help='')
    parser.add_option('-u', '--units', type='string', default='sec', help='')
    parser.add_option('-t', '--transport', type='string', default='car', help='')
    parser.add_option('-s', '--status', type='string', default='enabled', help='')
    
    (options, args) = parser.parse_args()

    # Validate required parameters
    if not options.api:
        parser.error('You have to choose API provider [\'SKOBBLER\', \'HERE\']')
    if options.api not in ['SKOBBLER', 'HERE']:
        parser.error('Invalid API provider [\'SKOBBLER\', \'HERE\']')
    if not options.key:
        parser.error('API KEY is required')
    if not options.file:
        parser.error('File is required (.csv with \',\' or \';\' as delimiter)')
    if not os.path.isfile(options.file):
        parser.error('File doesn\'t exist.')
    # Validate optional parameters
    if options.api == 'SKOBBLER':
        if options.units not in ['sec', 'meter']:
            parser.error('Invalid units type [\'sec\', \'meter\']')
        if options.transport not in ['pedestrian', 'bike', 'car']:
            parser.error('Invalid units type [\'pedestrian\', \'bike\', \'car\']')
    if options.api == 'HERE':
        try:
            options.app_id = options.key.split(',')[0]
            options.app_code = options.key.split(',')[1]
        except:
            parser.error('Wrong API key format')
        if not options.units or options.units not in ['time', 'distance']:
            options.units = 'time' 
        if options.units not in ['time', 'distance']:
            parser.error('Invalid units type [\'time\', \'distance\']')
        if options.transport not in ['pedestrian', 'car']:
            parser.error('Invalid units type [\'pedestrian\', \'car\']')
        if options.status not in ['enabled', 'disabled']:
            parser.error('Invalid traffic input [\'enabled\', \'disabled\']')
    
    print """Catchments params:\n
            api: {}
            key: {}
            points: {}
            range: {}
            units: {}
            transport: {}
            traffic status: {} (only for HERE)\n"""\
        .format(options.api, options.key, options.file, options.range,
                options.units, options.transport, options.status)

    # Read center points from file
    print 'Loading file with input points...'
    input_file = open(options.file)
    center_points = create_dict_reader(input_file)
    print 'Successful'
    # Get catchments (SKOBBLER or HERE) 
    if options.api == 'SKOBBLER':
        print 'Requesting catchments from SKOBBLER API...'
        for point in center_points:
            catchment = get_skobbler_catchment(point, options)
            geojson_feature = skobbler_catchment_to_geojson(catchment)
            file_name = save_as_geojson(geojson_feature, options)
            print '{} file successfully created'.format(file_name)
    else:
        print 'Requesting catchments from HERE API...'
        for point in center_points:
            catchment = get_here_catchment(point, options)
            geojson_feature = here_catchment_to_geojson(catchment)
            file_name = save_as_geojson(geojson_feature, options)
            print '{} file successfully created'.format(file_name)
    # Close input file
    input_file.close()
    return True
    

def create_dict_reader(file):
        dialect = csv.Sniffer().sniff(file.read(), delimiters=';,')
        file.seek(0)
        return csv.DictReader(file, dialect=dialect)


def get_skobbler_catchment(point, options):
    url = 'http://{0}.tor.skobbler.net/tor/RSngx/RealReach/json/20_5/en/{0}'\
        .format(options.key, options.key)
    params = {'start': '{1},{0}'.format(point['x'], point['y']),
              'transport': options.transport,
              'range': options.range,
              'units': options.units,
              'toll': 1,
              'highways': 1,
              'nonReachable': 0,
              'response_type': 'gps'}
    r = requests.get(url, params=params)
    catchment = json.loads(r.text)
    catchment['name'] = point['name']
    return catchment

def skobbler_catchment_to_geojson(catchment):
    geojson_feature = {"type": "Feature", "geometry": {
                    "type": "Polygon", "coordinates": [[]]},
                    "properties": {"name": catchment['name']}}
    coords = iter(catchment['realReach']['gpsPoints'][8:])
    for coord in coords:
        geojson_feature['geometry']['coordinates'][0].append([coord, next(coords)])
    geojson_feature['geometry']['coordinates'][0].append(
        [catchment['realReach']['gpsPoints'][8:][0],
        catchment['realReach']['gpsPoints'][8:][1]])
    return geojson_feature


def get_here_catchment(point, options):
    url = 'https://isoline.route.cit.api.here.com/routing/7.2/calculateisoline.json'
    params = {'start': 'geo!{1},{0}'.format(point['x'], point['y']),
              'mode': 'fastest;{};traffic:{}'.format(options.transport, options.status),
              'range': options.range,
              'rangetype': options.units,
              'app_id': options.app_id,
              'app_code': options.app_code}
    r = requests.get(url, params=params)
    catchment = json.loads(r.text)
    catchment['name'] = point['name']
    return catchment


def here_catchment_to_geojson(catchment):
    geojson_feature = {"type": "Feature",
                    "geometry": {"type": "Polygon", "coordinates": [[]]},
                    "properties": {"name": catchment['name']}}
    crds = []
    for row in catchment['response']['isoline'][0]['component'][0]['shape']:
        splt = row.split(',')
        crds.append(float(splt[1]))
        crds.append(float(splt[0]))
    coords = iter(crds)
    for coord in coords:
        geojson_feature['geometry']['coordinates'][0].append([coord, next(coords)])
    geojson_feature['geometry']['coordinates'][0].append([crds[0], crds[1]])
    return geojson_feature


def save_as_geojson(feature, options):
    file_name = feature['properties']['name'].lower().replace(' ', '_')
    file_suffix = '_{}_{}_{}_{}.geojson'\
        .format(options.transport, options.range, options.units, options.api)
    with open('{0}{1}'.format(file_name, file_suffix), 'w') as file_:
        file_.write(json.dumps(feature, indent=2))
    return file_name + file_suffix

      
if __name__ == '__main__':
    main()
